### PR TITLE
[TS migration] Migrate 'Tooltip' component to TypeScript

### PR DESCRIPTION
**Pull request by kosmydel**
### Fixed Issues

$ https://github.com/Expensify/App/issues/25123
PROPOSAL: N/A

### Tests

- [ ] Verify that no errors appear in the JS console
- [ ] On web and desktop:
  - [ ] Hover the buttons with tooltips (Search button, account settings button, FAB). Verify tooltips are correctly displayed.
- [ ] Verify that the app works.

### Offline tests

N/A

### QA Steps

- [ ] Verify that no errors appear in the JS console
- [ ] Tests from above.

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

Screenshot 2023-11-28 at 15 23 45

</details>

<details>
<summary>Android: mWeb Chrome</summary>

Screenshot 2023-11-28 at 15 23 45

</details>

<details>
<summary>iOS: Native</summary>

Simulator Screenshot - iPhone 15 Pro - 2023-11-28 at 15 24 25

</details>

<details>
<summary>iOS: mWeb Safari</summary>

Simulator Screenshot - iPhone 15 Pro - 2023-11-28 at 15 25 28

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/104823336/27d37649-47b0-47c6-9579-2091479a7401

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/104823336/d3bedf67-c906-43be-8637-2d9357bf259c

</details>

### CODE REVIEW

File: src/components/Tooltip/BaseTooltip.tsx
Comment (kosmydel) [inline src/components/Tooltip/BaseTooltip.tsx:198]: We have to add a new `result` function which returns a number instead of a string.

File: src/components/DisplayNames/DisplayNamesTooltipItem.tsx
Comment (blazejkustra) [inline src/components/DisplayNames/DisplayNamesTooltipItem.tsx:45]: Why was it changed to useMemo?

File: src/components/Pressable/PressableWithDelayToggle.tsx
Comment (blazejkustra) [inline src/components/Pressable/PressableWithDelayToggle.tsx:107]: Any clue if it's necessary?

File: src/components/Tooltip/BaseTooltip.native.tsx
Comment (blazejkustra) [inline src/components/Tooltip/BaseTooltip.native.tsx:10]:
```suggestion
import ChildrenProps from '@src/types/utils/ChildrenProps';

function Tooltip({children}: ChildrenProps) {
    return children;
}
```

File: src/components/Tooltip/BaseTooltip.tsx
Comment (blazejkustra) [inline src/components/Tooltip/BaseTooltip.tsx:2]: Let's get rid of lodash

File: src/components/Tooltip/BaseTooltip.tsx
Comment (blazejkustra) [inline src/components/Tooltip/BaseTooltip.tsx:46]: Did you try to use for of here?

File: src/components/Tooltip/types.ts
Comment (blazejkustra) [inline src/components/Tooltip/types.ts:43]:
```suggestion
export default TooltipProps;
export type {TooltipExtendedProps};
```

File: src/components/Tooltip/types.ts
Comment (blazejkustra) [inline src/components/Tooltip/types.ts:35]: Let's remove `tooltipRef` and instead use `forwardedRef` (second argument in a component).

File: src/components/Tooltip/types.ts
Comment (blazejkustra) [inline src/components/Tooltip/types.ts:12]: It should be `ReactNode`, or better use `ChildrenProps`:
```ts
import ChildrenProps from '@src/types/utils/ChildrenProps';
```

File: src/types/utils/callOrReturn.ts
Comment (blazejkustra) [inline src/types/utils/callOrReturn.ts:3]:
```suggestion
function callOrReturn<TValue>(value: TValue | (() => TValue)): TValue {
    if (typeof value === 'function') {
        return (value as () => TValue)();
    }

    return value;
}
```

File: src/styles/getTooltipStyles.ts
Comment (blazejkustra) [inline src/styles/getTooltipStyles.ts:150]: Please test it @kosmydel 😄

File: src/components/Tooltip/BaseTooltip.tsx
Comment (blazejkustra) [inline src/components/Tooltip/BaseTooltip.tsx:62]:
```suggestion
    renderTooltipContent,
```

File: src/components/Tooltip/TooltipRenderedOnPageBody.tsx
Comment (blazejkustra) [inline src/components/Tooltip/TooltipRenderedOnPageBody.tsx:49]:
```suggestion
    renderTooltipContent,
```

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: src/components/Tooltip/PopoverAnchorTooltip.tsx
Comment (BartoszGrajdek) [inline src/components/Tooltip/PopoverAnchorTooltip.tsx:13]: Maybe let's do it like this? You need to check if the logic is the same though
```suggestion
        const tooltipNode = tooltipRef.current?.['_childNode'] ?? null;
```

File: src/components/Tooltip/BaseTooltip.tsx
Comment (BartoszGrajdek) [inline src/components/Tooltip/BaseTooltip.tsx:205]: Why has this changed?

File: src/components/Tooltip/BaseTooltip.tsx
Comment (BartoszGrajdek) [inline src/components/Tooltip/BaseTooltip.tsx:95]: Can we do this? You should test it first though
```suggestion
    const updateTargetAndMousePosition = useCallback((e: MouseEvent) => {
        if (e.currentTarget instanceof HTMLElement) {
            target.current = e.currentTarget;
            initialMousePosition.current = {x: e.clientX, y: e.clientY};
        }
```

File: src/components/Tooltip/TooltipRenderedOnPageBody.tsx
Comment (blazejkustra) [inline src/components/Tooltip/TooltipRenderedOnPageBody.tsx:32]: Should be required?

File: src/components/Tooltip/TooltipRenderedOnPageBody.tsx
Comment (blazejkustra) [inline src/components/Tooltip/TooltipRenderedOnPageBody.tsx:36]: Should it be required?

**Review by blazejkustra — APPROVED** [review]

**Review by BartoszGrajdek — COMMENTED** [review]
Let's also move BaseTooltip files into a separate folder `src/components/Tooltip/BaseTooltip/....`

**Review by blazejkustra — APPROVED** [review]

**Review by BartoszGrajdek — APPROVED** [review]
LGTM!

File: src/components/Pressable/PressableWithDelayToggle.tsx
Comment (blazejkustra) [reply src/components/Pressable/PressableWithDelayToggle.tsx:106]: https://github.com/Expensify/App/pull/31544#discussion_r1405891178

File: src/components/Tooltip/BaseTooltip/index.tsx
Comment (kosmydel) [reply src/components/Tooltip/BaseTooltip/index.tsx:93]: Without it, there is this type of error:

Screenshot 2023-12-04 at 09 58 10

File: src/components/Tooltip/BaseTooltip/index.tsx
Comment (kosmydel) [reply src/components/Tooltip/BaseTooltip/index.tsx:204]: `Str.result` is only for strings, while `callOrReturn` is generic.

**Review by blazejkustra — APPROVED** [review]

**Review by fabioh8010 — APPROVED** [review]

**Comment by Santhosh-Sellavel** [comment]
Requested for new C+ plus, @allroundexperts will take over this.

**Comment by blazejkustra** [comment]
@kosmydel Please resolve conflicts 😄

**Comment by blazejkustra** [comment]
@allroundexperts is already responsible for two quite big TS migration PRs:
- [Report](https://github.com/Expensify/App/pull/28730)
- [AppNavigator](https://github.com/Expensify/App/pull/29157#issuecomment-1840852914)

(+ and this one)

Please take this into consideration before tackling all these at once 😅

**Comment by allroundexperts** [comment]
Have you listed these priority wise?

**Comment by blazejkustra** [comment]
@allroundexperts priority-wise:
1. AppNavigator
2. Tooltip
3. Report

**Comment by allroundexperts** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>


https://github.com/Expensify/App/assets/30054992/72cbbdd2-5d32-4822-9c48-bf66c863fb8f

</details>

<details>
<summary>Android: mWeb Chrome</summary>


https://github.com/Expensify/App/assets/30054992/494a71de-d1bf-4f2b-be40-aa6f60a3c30c

</details>

<details>
<summary>iOS: Native</summary>


https://github.com/Expensify/App/assets/30054992/45826261-fef3-4bd6-934e-2b30c69da9d2

</details>

<details>
<summary>iOS: mWeb Safari</summary>


Uploading Screen Recording 2023-12-06 at 4.57.16 AM.mov…

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>


https://github.com/Expensify/App/assets/30054992/372eb47a-6461-4957-9cd4-a93d7c5b3a39

</details>

<details>
<summary>MacOS: Desktop</summary>


https://github.com/Expensify/App/assets/30054992/1f470eb0-c5df-44c6-8758-41db030c99db

</details>

**Comment by allroundexperts** [comment]
Testing well. Waiting for the conflicts to be resolved before final approval.

**Comment by kosmydel** [comment]
Hey, I pulled the main, resolved conflicts, and had to make [fix warning on native](https://github.com/Expensify/App/pull/31544/commits/203cd81b0dd64b1d81661590c4058d4ff66b8c8c) on native, as there was a warning during runtime (related to `forwardRef`, web version used is while native didn't).

cc @blazejkustra, @allroundexperts

**Review by blazejkustra — APPROVED** [review]

**Review by allroundexperts — APPROVED** [review]
Approving since this is catching conflicts so fast. This is testing well and should be good to merge!

**Comment by kosmydel** [comment]
I've resolved conflicts, so it is ready to proceed.

**Comment by allroundexperts** [comment]
🎀 👀 🎀

**Review by allroundexperts — APPROVED** [review]

**Review by allroundexperts — APPROVED** [review]
Got an internal engineer!

**Comment by kosmydel** [comment]
I've resolved conflicts once again. cc @blazejkustra @allroundexperts

**Review by blazejkustra — APPROVED** [review]

**Review by allroundexperts — APPROVED** [review]

**Comment by mountiny** [comment]
Gonna take it over here as its component with lots of conflicts to get it over the finish line

File: src/components/Tooltip/types.ts
Comment (mountiny) [inline src/components/Tooltip/types.ts:30]: NAB
```suggestion
    /** Passes this down to Hoverable component to decide whether to handle the scroll behaviour to show hover once the scroll ends */
```

**Review by mountiny — APPROVED** [review]
Ok lets go, thank you for patience @kosmydel and testing thoroughly @allroundexperts

**Pull request merged**
